modules/service/src/main/resources/db/migration/V0881__program_calibration_role.sql

```diff
@@ -0,0 +1,247 @@
+            RAISE EXCEPTION '% programs must define an instrument', INITCAP(NEW.c_program_type);
+          ELSEIF (NEW.c_program_type != OLD.c_program_type)         OR
+                 (NEW.c_semester   IS DISTINCT FROM OLD.c_semester)   OR
+                 (NEW.c_instrument IS DISTINCT FROM OLD.c_instrument) THEN
+            NEW.c_semester_index := next_semester_index(NEW.c_program_type, NEW.c_semester, NEW.c_instrument);
+          END IF;
+          NEW.c_library_desc    := NULL;
+          NEW.c_science_subtype := NULL;
+        END;
+
+      WHEN NEW.c_program_type = 'example' THEN
+        BEGIN
+          IF NEW.c_instrument IS NULL THEN
+            RAISE EXCEPTION 'Example programs must define an instrument';
+          END IF;
+          NEW.c_semester        := NULL;
+          NEW.c_semester_index  := NULL;
+          NEW.c_library_desc    := NULL;
+          NEW.c_science_subtype := NULL;
+        END;
+
+      WHEN NEW.c_program_type = 'system' THEN
+        BEGIN
+          NEW.c_semester        := NULL;
+          NEW.c_semester_index  := NULL;
+          NEW.c_science_subtype := NULL;
+        END;
+
+      WHEN NEW.c_program_type = 'library' THEN
+        BEGIN
+          IF NEW.c_instrument IS NULL THEN
+            RAISE EXCEPTION 'Library programs must define an instrument';
+          ELSEIF NEW.c_library_desc IS NULL THEN
+            RAISE EXCEPTION 'Library programs must define a description';
+          END IF;
+          NEW.c_semester        := NULL;
+          NEW.c_semester_index  := NULL;
+          NEW.c_science_subtype := NULL;
+        END;
+
+      WHEN NEW.c_program_type = 'science' THEN
+        BEGIN
+          IF NEW.c_proposal_status <> 'not_submitted' THEN
+            IF NEW.c_semester IS NULL THEN
+              RAISE EXCEPTION 'Submitted science programs must define a semester';
+            ELSEIF NEW.c_science_subtype IS NULL THEN
+              RAISE EXCEPTION 'Submitted science programs must define a science subtype.';
+            ELSEIF (NEW.c_program_type != OLD.c_program_type) OR
+                   (NEW.c_semester IS DISTINCT FROM OLD.c_semester) OR
+                   (OLD.c_proposal_status = 'not_submitted' AND NEW.c_semester_index IS NULL) THEN
+              NEW.c_semester_index := next_semester_index('science', NEW.c_semester, NULL);
+            END IF;
+          END IF;
+          NEW.c_instrument   := NULL;
+          NEW.c_library_desc := NULL;
+        END;
+    END CASE;
+
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Calibration programs' targets are marked with the same calibration type
+CREATE OR REPLACE FUNCTION calibration_targets_on_calibration_programs()
+  RETURNS trigger AS $$
+DECLARE
+    calibration_role e_calibration_role;
+BEGIN
+    -- Fetch the value from the source table
+    SELECT c_calibration_role INTO calibration_role
+    FROM t_program
+    WHERE c_program_id = NEW.c_program_id;
+
+    NEW.c_calibration_role := calibration_role;
```

Comment:
What happens if the program calibration role changes after the program has been given one or more targets?  I'm a little unsure about keeping these in sync.  If we want to do it, we might consider making a `UNIQUE` constraint on the (`c_program_id`, `c_calibration_role`) pair in `t_program` and then having updating the `t_target` FK constraint to refer to them both.  That would at least prevent them from getting out of sync at the database level.

Reply:
The main thing I wanted to achieve is that if you add a target to a program with a calibration role they get in sync. Since there is no api to change the calibation role I thought it could be safe. You can change the role is SQL though

Reply:
If it cannot / should not change, we could enforce that in a trigger I guess.

Reply:
Added a trigger to forbid changing the calibration role